src/IB/IBFEMethod.cpp

```diff
@@ -227,15 +231,413 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
     return;
 }
 
-static const Real PENALTY = 1.e10;
+void
+assemble_cg_diffusion(EquationSystems& es, const std::string& system_name)
+{
+    const MeshBase& mesh = es.get_mesh();
+    const BoundaryInfo& boundary_info = *mesh.boundary_info;
+    const unsigned int dim = mesh.mesh_dimension();
+    auto& system = es.get_system<TransientLinearImplicitSystem>(IBFEMethod::PHI_SYSTEM_NAME);
+    const DofMap& dof_map = system.get_dof_map();
+    FEType fe_type = dof_map.variable_type(0);
+
+    UniquePtr<FEBase> fe(FEBase::build(dim, fe_type));
+    QGauss qrule(dim, FIFTH);
+    fe->attach_quadrature_rule(&qrule);
+    UniquePtr<FEBase> fe_face(FEBase::build(dim, fe_type));
+    QGauss qface(dim - 1, FIFTH);
+    fe_face->attach_quadrature_rule(&qface);
+
+    const std::vector<Real>& JxW = fe->get_JxW();
+    const std::vector<std::vector<Real> >& phi = fe->get_phi();
+    const std::vector<std::vector<RealGradient> >& dphi = fe->get_dphi();
+
+    const std::vector<std::vector<Real> >& phi_face = fe_face->get_phi();
+    const std::vector<Real>& JxW_face = fe_face->get_JxW();
+
+    DenseMatrix<Number> Ke;
+    std::vector<dof_id_type> dof_indices;
+
+    const Real dt = es.parameters.get<Real>("dt");
+    const Real PENALTY = es.parameters.get<Real>("cg_penalty");
+    const Real diffusion = es.parameters.get<Real>("Phi_diffusion");
+
+    MeshBase::const_element_iterator el = mesh.active_local_elements_begin();
+    const MeshBase::const_element_iterator end_el = mesh.active_local_elements_end();
+    for (; el != end_el; ++el)
+    {
+        const Elem* elem = *el;
+        dof_map.dof_indices(elem, dof_indices);
+        fe->reinit(elem);
+        unsigned int Ke_size = static_cast<unsigned int>(dof_indices.size());
+        Ke.resize(Ke_size, Ke_size);
+        for (unsigned int qp = 0; qp < qrule.n_points(); qp++)
+        {
+            for (unsigned int i = 0; i < phi.size(); i++)
+            {
+                for (unsigned int j = 0; j < phi.size(); j++)
+                {
+                    Ke(i, j) +=
+                        (phi[i][qp] * phi[j][qp] + diffusion * 0.5 * dt * (dphi[i][qp] * dphi[j][qp])) * JxW[qp];
+                }
+            }
+        }
+
+        for (unsigned int side = 0; side < elem->n_sides(); side++)
+        {
+            // penalty method to apply boundary conditions
+            if (is_physical_bdry(elem, side, boundary_info, dof_map))
+            {
+                fe_face->reinit(elem, side);
+                for (unsigned int qp = 0; qp < qface.n_points(); qp++)
+                {
+                    for (unsigned int i = 0; i < phi_face.size(); ++i)
+                    {
+                        for (unsigned int j = 0; j < phi_face.size(); ++j)
+                        {
+                            Ke(i, j) += PENALTY * phi_face[i][qp] * phi_face[j][qp] * JxW_face[qp];
```

Comment:
it would be a good thing to document how this works.

this penalty approach is related to a Robin boundary condition and using polynomial approximations which don't vanish on the boundary.  there are probably other ways to do this, but it works fine.